### PR TITLE
Migrated to null-safety

### PR DESCRIPTION
**Pull request by elias8**
I have migrated to null-safety. Also included a fix from #2.

### CODE REVIEW

**Comment by escamoteur** [comment]
Hi, 
sorry  but I did my own attempt on null safety because it's not done by just adding `?` or `!` if we want to leverage the advantages of null safety.
you find my approach here
https://github.com/escamoteur/functional_listener/tree/null-safety
I used the Dart Migration tool, but the result is a bit strange and some of the tests now fail. I will create an issue for it and reference it here. But still, thanks for your efforts.

**Comment by elias8** [comment]
Cool. But I actually tried to use the migration tool first and I found it misleading(especially on the use of the `late` keyword) sometimes and I migrated manually since the project was small. 

For instance, [here](https://github.com/escamoteur/functional_listener/blob/a775f4b18209b8764c3c7d4cb20535b64d421c89/lib/src/functional_value_notifiers.dart#L9)  I didn't find it good to use the `late` keyword because it is not guaranteed that the variable will be initialized(but still the migration tool suggests to use `late`). In our case, the subclass may or may not initialize it, but I think we should not be dependent on that. So, what I did was to make it nullable and check the value if it is null or not before using it.

I say using `late` and `!` could be dangerous since they won't show up a compile-time error but a runtime error. Just like using [this](https://github.com/escamoteur/functional_listener/blob/a775f4b18209b8764c3c7d4cb20535b64d421c89/lib/src/functional_value_notifiers.dart#L64) at [here](https://github.com/escamoteur/functional_listener/blob/a775f4b18209b8764c3c7d4cb20535b64d421c89/lib/src/functional_value_notifiers.dart#L73), the code successfully compiles but fails at runtime because of the use of the variable before initializing it (throws `LateInitializationError`). And that is because the compiler is not smart enough to know if a late variable is used after being initialized.

**Comment by escamoteur** [comment]
Actually I don't understand how it is possible that the ´Timer´ isn't initialized during the constructor. How is it possible that we get this exception?

**Comment by elias8** [comment]
The timer is not getting initialized because it is inside the `internalHandler` callback. But what is causing the exceptions is when we are trying to cancel the timer before it gets initialized.

```dart
internalHandler = () {
      debounceTimer.cancel(); // This is where the exception is thrown 
      debounceTimer =
          Timer(debounceDuration, () => value = previousInChain.value);
    };
```
By the way, I am not sure why we need to cancel the timer at the start.

**Comment by escamoteur** [comment]
Ah, thanks, I didn't realize that it's inside the handler. I really wished Dart would allow to write `=>{}` it would be way easier to read.

Yes we have to cancel the a running timer because otherwise it will assign it's value before the new timer completes.
I will change it to a nullable again.

**Comment by elias8** [comment]
But the timer is an instance variable, and we don't have access to the old-timer from the constructor. Or am I missing something? 🤔

**Comment by escamoteur** [comment]
yes we have, we capture the timer variable in in the lambda function.

**Comment by elias8** [comment]
Okay. I will take a look. Thanks!

File: lib/functional_listener.dart [hunk 1]
```diff
@@ -221,7 +216,9 @@ class ListenableSubscription {
   void cancel() {
     assert(handler != null);
     if (!canceled) {
-      endOfPipe.removeListener(handler);
+      if (handler != null) {
```
Comment (leafpetersen) [inline lib/functional_listener.dart:219]: Slightly odd to both assert that handler is not null but then explicitly handle the case that it's null.  Is this to be maximally defensive in production code?

Comment (escamoteur) [reply lib/functional_listener.dart:219]: Agreed, I close this PR here as on the null-safety branch is my current working state

Comment (escamoteur) [reply lib/functional_listener.dart:219]: @leafpetersen thanks for haveing a look at this

**Pull request closed without merging**